db/revtree.go

```diff
@@ -260,19 +260,23 @@ func (tree RevTree) findAncestorFromSet(revid string, ancestors []string) string
 }
 
 // Records a revision in a RevTree.
-func (tree RevTree) addRevision(info RevInfo) {
+func (tree RevTree) addRevision(info RevInfo) (err error) {
 	revid := info.ID
 	if revid == "" {
-		panic("empty revid is illegal")
+		err = errors.New("empty revid is illegal")
+		return
 	}
 	if tree.contains(revid) {
-		panic(fmt.Sprintf("already contains rev %q", revid))
+		err = errors.New(fmt.Sprintf("already contains rev %q", revid))
+		return
 	}
 	parent := info.Parent
 	if parent != "" && !tree.contains(parent) {
-		panic(fmt.Sprintf("parent id %q is missing", parent))
+		err = errors.New(fmt.Sprintf("parent id %q is missing", parent))
+		return
 	}
 	tree[revid] = &info
+	return nil
```

Comment:
Since RevTree doesn't know the docid, I think it's important for any callers that log/return the error message to add the docid (e.g. wrap the error with another error that includes doc id).  Without the docid it's difficult for users/support to dig into the issue.

As an alternative, you could add the docid as a parameter to addRevision just to get it into the error message.